fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -290,20 +289,62 @@ private void loadFromProto(@Nonnull RecordMetaDataProto.MetaData metaDataProto,
                                @Nonnull Descriptors.FileDescriptor[] dependencies,
                                boolean processExtensionOptions) {
         recordsDescriptor = buildFileDescriptor(metaDataProto.getRecords(), dependencies);
-        validateRecords(recordsDescriptor);
-        unionDescriptor = initRecordTypes(processExtensionOptions);
+        unionDescriptor = initRecordTypes(recordsDescriptor, processExtensionOptions);
         loadProtoExceptRecords(metaDataProto);
         processSchemaOptions(processExtensionOptions);
     }
 
     private void loadFromFileDescriptor(@Nonnull Descriptors.FileDescriptor fileDescriptor,
                                         boolean processExtensionOptions) {
         recordsDescriptor = fileDescriptor;
-        validateRecords(fileDescriptor);
-        unionDescriptor = initRecordTypes(processExtensionOptions);
+        unionDescriptor = initRecordTypes(fileDescriptor, processExtensionOptions);
         processSchemaOptions(processExtensionOptions);
     }
 
+    @Nonnull
+    private Descriptors.Descriptor initRecordTypes(@Nonnull Descriptors.FileDescriptor fileDescriptor, boolean processExtensionOptions) {
+        Descriptors.Descriptor unionDescriptor = fetchUnionDescriptor(fileDescriptor);
+        validateRecords(fileDescriptor, unionDescriptor);
+        fillUnionFields(unionDescriptor, processExtensionOptions);
+        return localUnionDescriptor == null ? unionDescriptor : localUnionDescriptor;
+    }
+
+    @Nonnull
+    private Descriptors.Descriptor fetchUnionDescriptor(@Nonnull Descriptors.FileDescriptor fileDescriptor) {
+        @Nullable Descriptors.Descriptor union = null;
+        for (Descriptors.Descriptor descriptor : fileDescriptor.getMessageTypes()) {
+            RecordMetaDataOptionsProto.RecordTypeOptions recordTypeOptions = descriptor.getOptions()
+                    .getExtension(RecordMetaDataOptionsProto.record);
+            if (recordTypeOptions != null && recordTypeOptions.hasUsage()) {
+                switch (recordTypeOptions.getUsage()) {
+                    case UNION:
+                        if (union != null) {
+                            throw new MetaDataException("Only one union descriptor is allowed");
+                        }
+                        union = descriptor;
+                        continue;
+                    case NESTED:
+                        continue;
```

Comment:
I think it's an accident of editing history that `NESTED` works differently from `RECORD`, etc.

Reply:
I kept it here intentionally because a NESTED message type cannot be a union descriptor and doesn't need the next check (at line 334). But now thinking thru it again, I guess it's missing a check that its name is not `DEFAULT_UNION_NAME`. Let me fix it.